sound/soc/sof/sof-client.c

```diff
@@ -216,6 +216,52 @@ int sof_client_ipc_tx_message(struct sof_client_dev *cdev, void *ipc_msg,
 }
 EXPORT_SYMBOL_NS_GPL(sof_client_ipc_tx_message, SND_SOC_SOF_CLIENT);
 
+int sof_suspend_clients(struct snd_sof_dev *sdev, pm_message_t state)
+{
+	struct auxiliary_driver *adrv;
+	struct sof_client_dev *cdev;
+
+	mutex_lock(&sdev->ipc_client_mutex);
+
+	list_for_each_entry(cdev, &sdev->ipc_client_list, list) {
+		/* Skip devices without loaded driver */
+		if (!cdev->auxdev.dev.driver)
+			continue;
+
+		adrv = to_auxiliary_drv(cdev->auxdev.dev.driver);
+		if (adrv->suspend)
+			adrv->suspend(&cdev->auxdev, state);
+	}
+
+	mutex_unlock(&sdev->ipc_client_mutex);
+
+	return 0;
+}
+EXPORT_SYMBOL_NS_GPL(sof_suspend_clients, SND_SOC_SOF_CLIENT);
+
+int sof_resume_clients(struct snd_sof_dev *sdev)
+{
+	struct auxiliary_driver *adrv;
+	struct sof_client_dev *cdev;
+
+	mutex_lock(&sdev->ipc_client_mutex);
+
+	list_for_each_entry(cdev, &sdev->ipc_client_list, list) {
+		/* Skip devices without loaded driver */
+		if (!cdev->auxdev.dev.driver)
+			continue;
+
+		adrv = to_auxiliary_drv(cdev->auxdev.dev.driver);
+		if (adrv->resume)
+			adrv->resume(&cdev->auxdev);
```

Comment:
same, don't use the PM callback directly, ask the PM framework to resume the device for you.

Reply:
@ujfalusi why do we need to suppose suspend/resume for the trace client at all? The only time trace sends IPC's is during init/free which are done during probe and remove and the core will always be active at that point no? 

I dont think suspend/resuming clients from the core looks right at all

Reply:
we need to send IPC when the DSP resumes also.

I can not find other ways to to handle the dma-trace feature of the SOF. It is a passive feature, it can not enforce PM state, it must follow the core and act on the change.

Do you see any other way to handle a feature of SOF which should be following and not dictating (or enforcing) power state for the DSP?
I think I have covered most of the points in my comment above yours for @plbossart and @lyakh, did I missed something?

Reply:
What IPC does the trace client need to send for DSP resume? The core needs to send the CTX_RESTORE when the DSp is resumed. But I cant imagine the client having to send anything

Reply:
Please help me understand, Peter. why does the trace client need PM. It's job is just to do debugfs read/write isnt it? and a write is performed when an IPC arrives from the DSP in which case the DSP is already in D0. The core simply needs to forward the ipC RX to the trace client.  In the case of read, we dont really need the DSP to be in D0 right?

Reply:
The client conversion of dma-trace follows how the current code is written.
in `sound/soc/sof/pm.c` (https://github.com/thesofproject/linux/blob/d54b78032d387ac880cb426dfea388e9dbe158ba/sound/soc/sof/pm.c#L150) we have:
```
/* resume DMA trace, only need send ipc */
ret = snd_sof_init_trace_ipc(sdev);
if (ret < 0) {
	/* non fatal */
	dev_warn(sdev->dev,
			"warning: failed to init trace after resume %d\n",
			ret);
}
```
And indeed, the `snd_sof_init_trace_ipc()` sends IPC.
If it is not needed then it must be removed (and rigorously tested) prior to client conversion.

The `struct auxiliary_driver` have the `suspend` / `resume` callbacks since the initial commit, no extended information available, only this:
`The bus supports probe/remove shutdown and suspend/resume callbacks.`
I'm certain that they are added for a reason and I would not be surprised if the reason is for exactly the same use case as we are using them.

let's theorize: we want to have a simple client driver which would light up an LED when the DSP is powered on. Nothing more, nothing less. How it can be done?

Reply:
isn't this the case of a 'no callback device' described in https://www.kernel.org/doc/html/latest/power/runtime_pm.html?highlight=pm_runtime#no-callback-devices

I think you need some sort of signaling between the SOF core and the DMA trace or your LED thingy - using a loop on resume doesn't seem quite right here.

Reply:
@plbossart, yes it is and the dtrace client sets `pm_runtime_no_callbacks()`

I would love to use `device_for_each_child()` instead of our list, but sof_dev have other children than the clients.

There must be a reason why the auxdev have suspend and resume callbacks defined and I can not think of other use cases for them than this one.. In other places where the auxiliary bus is used have only one child so they obviously not doing any looping, but we are pushing this to it's limits. imho.